### DISCUSSION TITLE
Unable to disable 302 redirects

### DISCUSSION THREAD

**Discussion by gryffus** [nextcloud/all-in-one#1393] [Questions] [answered]
### Steps to reproduce
1. Use docker compose from https://github.com/nextcloud/all-in-one/discussions/575 (using Caddy proxy)
2. docker compose up -d
3. Use Caddyfile with multiple FQDNs proxying the requests to Nextcloud: 
```
https://fqdn1.nextcloud.com:443 {
    reverse_proxy localhost:11000
}
https://fqdn2.nextcloud.com:443 {
    reverse_proxy localhost:11000
}
https://fqdn3.nextcloud.com:443 {
    reverse_proxy localhost:11000
}
```
4. Modify config.php and remove "overwritehost" and any other "overwrite*" parameters

### Expected behavior
After visiting https://fqdn2.nextcloud.com, user should not be redirected to https://fqdn1.nextcloud.com

### Actual behavior
After visiting https://fqdn2.nextcloud.com, user is redirected to https://fqdn1.nextcloud.com via HTTP 302 redirect

### Host OS
Debian Linux

#### Nextcloud AIO version
2.1.2

#### Current channel
latest

**Comment by szaimen** [comment, marked as answer]
One AIO instance is indeed not meant to be used together with multiple domains. However you can follow https://github.com/nextcloud/all-in-one/blob/main/multiple-instances.md to create mutiple AIO instances, one for each domain :)

**Reply by gryffus** [reply]
Perhaps anyone has a working proxy config, that sends requests to the main AIO FQDN instead of the requested domain and blocks the redirects sent to clients? There must be a way to trick AIO into thinking it is serving one FQDN, while proxy serves it under multiple FQDNs. This is a very simple and common networking setup.

**Reply by gryffus** [reply]
Please note, that i can do SSL termination on the proxy, i do not require AIO to talk via SSL. Perhaps that should help accomplish it?

**Reply by gryffus** [reply]
Having separate instance of the whole stack, just for sake of serving under multiple DNS CNAMEs is a performance and maintenance nightmare. I just need to disable the 302 redirects Nextcloud is sending.

**Reply by szaimen** [reply]
AIO is simply not designed to be used in such a way. You will see what I mean if you manage to disable the redirect: the redirect is only one obstacle of many.

**Reply by gryffus** [reply]
Got it solved. Please do not mark your reply as an answer, the solution is shared in my own reply below. Thanks.

**Comment by gryffus** [comment]
Got it working.

Nextcloud devs apparently refuse to modify the redirect behaviour, so you need to solve it on proxy:

Here is a working example for Caddy proxy:

```
https://fqdn1.nextcloud.com:443 {
    reverse_proxy localhost:11000
}
https://fqdn2.nextcloud.com:443 {
    reverse_proxy localhost:11000 {
        header_up Host fqdn1.nextcloud.com
        header_down Location fqdn1.nextcloud.com fqdn2.nextcloud.com
    }
}
https://fqdn3.nextcloud.com:443 {
    reverse_proxy localhost:11000 {
        header_up Host fqdn1.nextcloud.com
        header_down Location fqdn1.nextcloud.com fqdn3.nextcloud.com
    }
}
```

Trick is to rewrite "Location" response header to the CNAME / FQDN you are using in current connection and to rewrite "Host" request header to the main FQDN of the nextcloud server. Works like a charm.

**Reply by Zoey2936** [reply]
I am not sure, but I think doing it this way creates security issues.

**Reply by gryffus** [reply]
> I am not sure, but I think doing it this way creates security issues.

Could you please elaborate? I can overcome any shortcommings on our web-app firewall. Currently we let only requests for the correct domain in, but we can extend the rules to a header values or something.

**Reply by Zoey2936** [reply]
Because every request to fqdn2.nextcloud.com and fqdn3.nextcloud.com will get excepted, without redirect, even by a non-authorized reverse proxy, which is proxing to fqdn2.nextcloud.com or fqdn3.nextcloud.com, but I'm not sure

**Reply by gryffus** [reply]
We are firewalling requests from the proxy to the backend, so only the proxy is able to reach actual NC instance.
On the internet-facing side from proxy, we have web application firewall that won't let requests for anything else than fqdn1.nextcloud.com, fqdn2.nextcloud.com and fqdn3.nextcloud.com. We are just outsourcing the proxy validation from the NC to the proxy itself (and firewall integrated to it), so i think it should be fine.

However, I will still talk about this concern with our security team and share the outcome here.

Thanks for pointing that out!

**Reply by gryffus** [reply]
We got a green from our security team, so i guess it is OK from security standpoint
